src/utils/heading-handler.js

```diff
@@ -55,7 +55,7 @@ class HeadingHandler {
       // Inject the id after transformation
       const n = Object.assign({}, node);
       const el = fallback(h, n);
-      el.properties.id = headingIdentifier;
+      el.properties.id = el.properties.id || `user-content-${headingIdentifier}`;
```

Comment:
If an `id` was provided, keep it as is, otherwise make sure to escape it to avoid DOM clobbering.
If provided `id` does clobber the DOM, DOMPurify will remove it during the sanitization step.

Reply:
The `user-content-` prefix you are inserting here seems odd and is leading to heading IDs that don't match GitHub's.

Reply:
Github does the same. You can check the helix home readme for instance and you'll notice that
```markdown
# Background
```
becomes
```html
<h2>
    <a id="user-content-background" class="anchor" aria-hidden="true" href="#background"></a>
    Background
</h2>
```

Reply:
I didn't know that the actual IDs in GitHub are different and that GitHub is using JavaScript to restore the scroll functionality they've broken through this clobbering.

I think we should not handle it the same way. Use proper `id` attribute values, and maybe add a blacklist of known DOM properties that are forbidden. You have access to JSDOM, which could help.

Reply:
It's a tricky one to handle, as white/black listing is not necessarily enough, you also have to account for all character encoding possibilities. Any value put in an `id` or `name` attribute can potentially clobber the DOM, and browsers are _supposed_ to prevent that (Chrome does, Firefox has a long history of not doing it, …)

If you only blacklist, then you fail to escape new properties that are introduced by the browser engine. The accepted norm here is to prefix using `user-content-` like github does it.

Reply:
By the way, if I let these through then DOMPurify just completely removes the identifiers, but then you can't target them in the URL either. In most cases, this is maybe not a problem, but the moment you start documenting APIs on a public website, chances are high that you use property names for the headings, and then you mostly end up with no heading identifiers at all.

Reply:
The potential attack vector is relatively small, as it would require
- a new DOM property
- that is not yet reflected in JSDOM
- that does not use camel-case syntax (our slugging makes everything lowercase)

Maybe enabling this prefix for single-word headlines only would be sufficient, as JS does not allow properties with a dash in them.

Reply:
Seems to me that the cure is worse than the disease.

Reply:
> Seems to me that the cure is worse than the disease.

Seems to me the this is a subjective argument and we could argue for hours on that ;)
Security experts seem to think otherwise if prefixing has been accepted as the best practice.

I'd rather have a security expert take the decision on that one. I'm not familiar enough with the topic to be able to judge what all the possible edge cases are for DOM Clobbering. Various articles I found seem to show it's quite easy to write them, so I'd be leaning towards the safest approach, that a custom one that we do not fully understand.

Also, in my previous PR on the headings, the ask was to behave exactly like github does, and here your ask is to not do so… I'm a bit confused.

Reply:
> Also, in my previous PR on the headings, the ask was to behave exactly like github does, and here your ask is to not do so… I'm a bit confused.

I was mistaken. I assumed that GitHub used the IDs that appeared when you hovered over the link symbol in a headline and took that for face value instead of inspecting the HTML.

GitHub's approach would require loading JavaScript into every customer site, just to get links to fragment identifiers to work as expected. This is not acceptable.

It looks like at least three static site generators I've looked at decided that this is not a problem:
- https://gohugo.io/templates/lookup-order/#hugo-layouts-lookup-rules
- https://jekyllrb.com/docs/#instructions
- https://www.gatsbyjs.org/tutorial/#advanced-tutorials

(note the fragment identifiers in the URL, which all work, and are all in the HTML, and are all not dangerous.)

Reply:
> GitHub's approach would require loading JavaScript into every customer site

Not necessarily. You can also achieve this by overriding the heading output and injecting the anchors and accepting that your has includes the prefix.
At the moment we only output the identifier, not the anchor to target it, so any site will have to add some logic on top. Whether it's clientisde JS or using a matcher function to extend the heading markup is probably a project decision.

> It looks like at least three static site generators I've looked at decided that this is not a problem:

I'd have to write some example on their stack to see how they really handle the issue:
- do they ignore it altogether?
- do they just remove the offending id/name like DOMPurify does?
- do they have another method?

The decision is also ultimately dependent on who the producer of the markup is?
DOM clobbering mostly requires a conscious decision to "hack" the content, so if we only look at authors writing content, the risk is mostly limited. But the moment you have a use case where the community is creating content via comments for the site, then you need more security.

I'd be fine ignoring the problem for now (as DOMPurify anyway cleans the id/name), and maybe documenting the limitation somewhere, if we only consider controlled authors to be the content generators.

Reply:
The markup producer is the author and trusted. I'd assume that most DOM clobbering issues that arise would be accidents rather than attacks.

Reply:
Ok, so assuming an "accident":
```
## Name
```

The process will look like this:
- generate a heading with identifier `name`
- DOM gets sanitized by DOMPurify which detects a collision on `document.name`, and removes it
- final output is:
```
<h2>Name</h2>
```
with no anchor.

If customers want the prefixed version, they can do that in their own matching function for the heading

Reply:
Looks good. This is assuming that

```markdown
## Name 1
```

would get an `id`, and `DOMPurify` isn't simply removing all `id` attributes.

Reply:
Yes,
```markdown
## Name 1
```
would be kept and turned into:
```html
<h2 id="name-1">Name 1</h2>
```

In the end, I only see this being a real issue on API documentation websites where the author would probably use the API name for the heading:
```markdown
# My Custom Object
## Properties
### Name
### OnClick
### OnChange
…
```

And for these edge cases, I would either:
- Add a prefix to all headings: `Property: Name`, `Event: OnClick` in the markdown
- Use a custom matcher function that injects a prefix in the identifier in the HTML
- Use a custom JS script that does it clientside

@trieloff, @kptdobe, @rofe:
Since this would be a known edge case, where should we document this?
Is helix using ADRs (https://adr.github.io/) or do we have customer-facing documentation that would need an entry for this?

Reply:
We don't use ADRs, but it looks like a good template for this, so just put the first ADR into docs. Removing the ID for `properties`, `name`, etc. only seems like a good solution and there are enough documented workarounds that customers can implement if needed.